src/commonMain/kotlin/arrow.exact/ExactDsl.kt

```diff
@@ -1,30 +1,26 @@
 package arrow.exact
 
-import arrow.core.*
+import arrow.core.Either
+import arrow.core.raise.Raise
+import arrow.core.raise.either
 
-internal class AndExact<A, B, C>(
-  private val exact1: Exact<A, B>,
-  private val exact2: Exact<B, C>
-) : Exact<A, C> {
+@DslMarker
+annotation class ExactDsl
 
-  override fun from(value: A): Either<ExactError, C> {
-    return exact1.from(value)
-      .flatMap { exact2.from(it) }
-  }
+@ExactDsl
+fun <A, R : Refined<A>> exact(
+  construct: Raise<ExactError>.(A) -> R
+): Exact<A, R> = object : Exact<A, R> {
+  override fun from(value: A): Either<ExactError, R> = either { construct(value) }
 }
 
-fun <A, B> exact(predicate: Predicate<A>, constructor: (A) -> B): Exact<A, B> {
-  return object : Exact<A, B> {
-    override fun from(value: A): Either<ExactError, B> {
-      return if (predicate.invoke(value)) {
-        constructor.invoke(value).right()
-      } else {
-        ExactError("Value ($value) doesn't match the predicate").left()
-      }
-    }
-  }
+@ExactDsl
+fun <E : Any, A, R : Refined<A>> exactEither(
+  construct: Raise<E>.(A) -> R
+): ExactEither<E, A, R> = object : ExactEither<E, A, R> {
+  override fun from(value: A): Either<E, R> = either { construct(value) }
 }
 
-infix fun <A, B, C> Exact<A, B>.and(other: Exact<B, C>): Exact<A, C> {
-  return AndExact(this, other)
-}
+// TODO: Add any relevant extensions to Refined<A>
+
+fun <A, B> Refined<A>.map(f: (A) -> B): B = f(value)
```

Comment:
So actually if I think about this.. Is `Refined<A>` not the same as `Either<ExactError, A>`? 🤔 And do we not get all required functions through that?

Reply:
Indeed, we get everything we need from `Either` - can't think of anything that we need in `Refined` but is missing in `Either`.